src/ui_fsmenu/singleplayerdropdown.cc

```diff
@@ -0,0 +1,384 @@
+/*
+ * Copyright (C) 2002-2020 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+
+#include "ui_fsmenu/singleplayerdropdown.h"
+
+#include <memory>
+
+#include <boost/algorithm/string.hpp>
+
+#include "ai/computer_player.h"
+#include "base/i18n.h"
+#include "graphic/playercolor.h"
+#include "map_io/map_loader.h"
+
+#define AI_NAME_PREFIX "ai" AI_NAME_SEPARATOR
+#define HUMAN_PLAYER "human_player"
+#define RANDOM "random"
```

Comment:
We have started replacing macros with constexpr, because this will give us a datatype too. Like this:

`constexpr const char* const kRandom = "random";`

Reply:
hmm, I tried that, see latest commit. But since I need to concat these strings, I am not sure if this is feasible. Internet says I'd need to use strcat, see https://stackoverflow.com/questions/1995053/const-char-concatenation/44204042

Please advise :)

Reply:
yeah, I think I did that already... ;)